### ISSUE TITLE
For the same complex dtype and same value, comparing a PyTorch tensor with a NumPy array results in False

### ISSUE THREAD

**Issue by xuhdev** [pytorch/pytorch#43579]
## 🐛 Bug

For the same complex dtype and same value, comparing a PyTorch tensor with a NumPy array results in False. Note [5] and [10] in the example below. The issue is gone if `complex64` is replaced with `float32` (i.e., both [5] and [10] output True for `float32`).

```python
In [1]: import numpy as np

In [2]: import torch

In [3]: a=torch.tensor([1, 2], dtype=torch.complex64)

In [4]: c=np.array([1, 2], dtype=np.complex64)

In [5]: a[0] == c[0]
Out[5]: False

In [6]: a[0].item() == c[0]
Out[6]: True

In [7]: a[0] == c[0].item()
Out[7]: tensor(True)

In [8]: a[0].item() == c[0].item()
Out[8]: True

In [9]: a=torch.from_numpy(c)

In [10]: a[0] == c[0]
Out[10]: False
```

## Environment

PyTorch version: 1.7.0.dev20200819+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: 7.0.1-8 (tags/RELEASE_701/final)
CMake version: version 3.16.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Quadro P400
Nvidia driver version: 440.100
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.7.0.dev20200819+cu101
[conda] Could not collect

**Comment by mruberry** [comment]
Thanks for reporting this issue, @xuhdev! I think we would take a fix for this.

**Comment by anjali411** [comment]
good catch @xuhdev is_numpy_scalar method [here] (https://github.com/pytorch/pytorch/blob/2b70f827373fe5c90e15ce7ba0a1d331dcd57379/torch/csrc/utils/tensor_numpy.cpp#L249) (called in CheckScalar) doesn't check for complex values which might be an issue. We should fix it regardless. 

```
>>> a1=torch.tensor([1, 2], dtype=torch.float32)
>>> c1=np.array([1, 2], dtype=np.float32)
>>> type(a1)
<class 'torch.Tensor'>
>>> type(c1)
<class 'numpy.ndarray'>
>>> a1[0]
tensor(1.)
>>> c1[0]
1.0
>>> a1[0] == c1[0]
tensor(True)
```

```
>>> a=torch.tensor([1, 2], dtype=torch.complex64)
>>> c=np.array([1, 2], dtype=np.complex64)
>>> type(a[0])
<class 'torch.Tensor'>
>>> type(c[0])
<class 'numpy.complex64'>
>>> a[0] == c[0]
False
```

Also, in the case of float32, the types of element compared are `'numpy.ndarray'` and `'torch.Tensor'`. However, in the case of complex 64, the type of elements compared are `'numpy.complex64'` and `'torch.Tensor'` which might not be properly handled.

**Comment by anjali411** [comment]
good catch @xuhdev is_numpy_scalar method [here] (https://github.com/pytorch/pytorch/blob/2b70f827373fe5c90e15ce7ba0a1d331dcd57379/torch/csrc/utils/tensor_numpy.cpp#L249) (called in CheckScalar) doesn't check for complex values which might be an issue. We should fix it regardless though.

**Comment by xuhdev** [comment]
@anjali411 Thanks, I've made #43644 to address this. However, the following case remains unfixed:

```python
a = torch.tensor([1 + 1j, 2 + 2j], dtype=torch.complex64)
c = np.array([1 + 1j, 2 + 2j], dtype=np.complex64)
print(a[0] == c[0])
```

output:

```
/home/user/src/pytorch/torch/tensor.py:25: ComplexWarning: Casting complex values to real discards the imaginary part return f(*args, **kwargs)
tensor(False)
```

**Comment by mruberry** [comment]
> @anjali411 Thanks, I've made #43644 to address this. However, the following case remains unfixed:
> 
> ```python
> a = torch.tensor([1 + 1j, 2 + 2j], dtype=torch.complex64)
> c = np.array([1 + 1j, 2 + 2j], dtype=np.complex64)
> print(a[0] == c[0])
> ```
> 
> output:
> 
> ```
> /home/user/src/pytorch/torch/tensor.py:25: ComplexWarning: Casting complex values to real discards the imaginary part return f(*args, **kwargs)
> tensor(False)
> ```

Good catch, @xuhdev. We'd definitely take a fix for that, too.

**Comment by anjali411** [comment]
@xuhdev I just checked and we still have this issue. would you like to work on a fix for it?

**Comment by xuhdev** [comment]
@anjali411 Yes, I can look into this, but probably in a couple of weeks. I've assigned myself so I won't forget.

**Comment by xuhdev** [comment]
@anjali411 I've narrowed the issue down to the extent that in https://github.com/pytorch/pytorch/issues/43579#issuecomment-681072392

> ComplexWarning: Casting complex values to real discards the imaginary part return f(*args, **kwargs)

is emitted by NumPy, not PyTorch. In addition, this is caused by `a[0].__eq__(c[0])`, not `c[0].__eq__(a[0])`. However, I still have a hard time finding where `__eq__` is implemented. Could you offer some help?

**Comment by anjali411** [comment]
@xuhdev it ultimately calls `at::eq`. I did some further digging and found out that `x[0] + y[0]` does not match `(x+y)[0]`. `x+y` produces the correct result so it's probably the conversion to `Scalar` that's causing the issue. The value stored in the scalar incorrectly has it's imaginary value set to 0. 
```
>>> import torch
>>> x = torch.randn(4, dtype=torch.cfloat)
>>> y = x.numpy()
>>> x[0] == y[0]
<stdin>:1: ComplexWarning: Casting complex values to real discards the imaginary part
self.real: -0.733001
[ CPUFloatType{} ]
self.imag: -0.872041
[ CPUFloatType{} ]
scalar.real: -0.733001
scalar.imag: 0
tensor(False)
>>> x[0] + y[0]
tensor(-1.4660-0.8720j)
>>> x
tensor([-0.7330-0.8720j,  0.8876+0.2475j, -0.3884+0.9800j,  0.6060+0.9870j])
>>> x+y
tensor([-1.4660-1.7441j,  1.7753+0.4950j, -0.7769+1.9600j,  1.2119+1.9741j])
>>> (x+y)[0]
tensor(-1.4660-1.7441j)
>>> x[0] + x[0]
tensor(-1.4660-1.7441j)
```

**Comment by kcarnold** [comment]
For the Torch version on Google Colab (`2.1.0+cu121`), this happens even with plain int/float types:

```python
>>> a = torch.tensor([1, 2])
>>> b = np.array([1, 2])
>>> a == b
False
```

[Colab notebook](https://colab.research.google.com/drive/1b9IlrhI7AF7KnOGwXI9XnYWv6ZJ4CEdN?usp=sharing).
